src/core/lwt.mli

```diff
@@ -21,502 +22,1785 @@
+| true -> Lwt_io.(read_line stdin)   (* Has type string Lwt.t... *)
+| false -> Lwt.return ""             (* ...so wrap empty string in a promise. *)
+]}
+
+    Another typical usage is in {{: #VALbind} [let%lwt]}. The expression after
+    the “[in]” has to evaluate to a promise. So, if you compute an ordinary
+    value instead, you have to wrap it:
+
+{[
+let%lwt line = Lwt_io.(read_line stdin) in
+Lwt.return (line ^ ".")
+]} *)
+
+val fail : exn -> _ t
+(** [Lwt.fail exn] is like {!Lwt.return}, except the new {{: #TYPEt} promise}
+    that is {e already rejected} with [exn]. *)
 
-val fail : exn -> 'a t
-  (** [fail e] is a thread that fails with the exception [e]. *)
 
-val fail_with : string -> 'a t
-  (** [fail_with msg] is a thread that fails with the exception
-      [Failure msg]. *)
 
-val fail_invalid_arg : string -> 'a t
-  (** [fail_invalid_arg msg] is a thread that fails with the exception
-      [Invalid_argument msg]. *)
+(** {3 Callbacks} *)
 
 val bind : 'a t -> ('a -> 'b t) -> 'b t
-  (** [bind t f] is a thread which first waits for the thread [t] to
-      terminate and then, if the thread succeeds, behaves as the
-      application of function [f] to the return value of [t].  If the
-      thread [t] fails, [bind t f] also fails, with the same
+(** [Lwt.bind p_1 f] makes it so that [f] will run when [p_1] is {{: #TYPEt}
+    {e fulfilled}}.
+
+    The value that [p_1] is fulfilled with is passed to [f]. [f] then returns
+    another promise, [p_2]. This second promise means [f] can spawn more I/O
+    operations – they will create [p_2].
+
+    [Lwt.bind] itself returns immediately. It only attaches the callback [f] to
+    [p_1] – it does not wait for [p_2]. {e What} [Lwt.bind] returns is yet a
+    third promise, [p_3]. Roughly speaking, fulfillment of [p_3] represents both
+    [p_1] and [p_2] becoming fulfilled, one after the other.
+
+    A minimal example of this is an echo program:
+
+{[
+let () =
+  let p_3 =
+    Lwt.bind
+      Lwt_io.(read_line stdin)
+      (fun line -> Lwt_io.printl line)
+  in
+  Lwt_main.run p_3
+
+(* ocamlfind opt -linkpkg -package lwt.unix code.ml && ./a.out *)
+]}
+
+    Rejection of [p_1] and [p_2], and raising an exception in [f], are all
+    forwarded to rejection of [p_3].
+
+    The rest is (1) a precise description, and (2) notes on why [Lwt.bind] is
+    always written using syntactic sugar.
+
+    {b (1)} [Lwt.bind] returns a promise [p_3] immediately. [p_3] starts out
+    pending, and is resolved as follows:
+
+    - The first condition to wait for is that [p_1] becomes resolved. It does
+      not matter whether [p_1] is already resolved when [Lwt.bind] is called, or
+      becomes resolved later – the rest of the behavior is the same.
+    - If and when [p_1] becomes resolved, it will, by definition, be either
+      fulfilled or rejected. Fulfilling is the interesting case, but rejection
+      is handled very simply, so we cover rejection of [p_1] first.
```

Comment:
This sentence about interesting and simple cases is not needed. (It's singposting but on an hyperlocal scale.)